### PR TITLE
Conner

### PR DESCRIPTION
**Pull request by cnnrdltn**
This is a big one

- added microphone support in AudioController
- Fixed AudioProcessor to correctly deal with times (sync up the input with the sheet music)
- Had to wrap the SeeScoire Play Data in a wrapper and add the SeeScoreNote structure to get objective e cpp to work with swift
- Note Coloring

### CODE REVIEW

File: Ivory/AudioProcessor.swift [hunk 1]
```diff
@@ -61,44 +61,36 @@ class AudioProcessor: NSObject {
     // check out the NoteIntervalMap protocol.  it's just opaque, audiocontroller
     // will use different backend implementations depending on input type most likely
     // currently everything is using midi version which works fine bc of Note(midiValue:value)
-    func checkValue(atTime time:Double, processedResults:Zip2<[Float],[Float]>) -> [Note]? {
-        // dummy, sample implementation
+    func checkValue(atTime time:Double, processedResults:Zip2<[Float],[Float]>) -> [SeeScoreNote:ETNoteEvaluationResultType]? {
+        var out:[SeeScoreNote:ETNoteEvaluationResultType] = [:]
         if let noteMap = self.noteIntervalMap {
-            // currently this method has a dummy implementation. hardcode expected values for now
-            println("expectedNotes at time \(time)")
```
Comment (cnnrdltn) [inline Ivory/AudioProcessor.swift]: document why *1000.  add "_ms" to variable to match seeScore conventions

File: Ivory/AudioProcessor.swift [hunk 2]
```diff
@@ -61,44 +61,36 @@ class AudioProcessor: NSObject {
...
+            println("expectedNotes at time \(time*1000)")
             // done in ms
-            println(noteMap.expectedNotesAt(Double(time), duration: Double(500)))
+            println(noteMap.expectedNotesAt(Double(time*1000), duration: Double(500)))
             var expectedNotes = noteMap.expectedNotesAt(Double(time*1000), duration: Double(500))
-
```
Comment (cnnrdltn) [inline Ivory/AudioProcessor.swift]: clear this dead code.

File: Ivory/AudioProcessor.swift [hunk 3]
```diff
@@ -159,15 +151,15 @@ class AudioProcessor: NSObject {
...
-         return fftResultFrequenciesTups
+         return checkedResults
+//         return fftResultFrequenciesTups
     }
 }
 
 
 func predictedNotes(fftResultFrequenciesTups: Zip2<[Float],[Float]>) -> [Pitch]?{
     // extract dominant frequencies from amplitudes array
-    let mininumViableAmplitude:Float = 0.000010
```
Comment (cnnrdltn) [inline Ivory/AudioProcessor.swift]: this is wrong

File: Ivory/AudioProcessor.swift (hunk 1 above)
Comment (fkmasuda) [inline Ivory/AudioProcessor.swift]: why is the Sample size 16384?

File: Ivory/IvoryKit/IvoryKit/AudioController.swift [hunk 4]
```diff
@@ -145,18 +166,76 @@ class AudioController:NSObject {
...
+    func handleTapData(audioProcessor:AudioProcessor, buf:AVAudioPCMBuffer,when:AVAudioTime) -> (){
+        if self.startTime == nil {
+            self.startTime = AVAudioTime.secondsForHostTime(when.hostTime)
+        }
+        // time may be offset based on current time.  only want the difference from the starting time for the song
+        var time = AVAudioTime.secondsForHostTime(when.hostTime)
+        time -= self.startTime!
+        
+        let channels = buf.floatChannelData
+        var floats = [Float](UnsafeBufferPointer(start: channels[0], count: Int(buf.frameLength)))
```
Comment (fkmasuda) [inline Ivory/IvoryKit/IvoryKit/AudioController.swift]: UnsafeBufferPointer? Unclear what this does

File: Ivory/SSSampleViewController.mm [hunk 5]
```diff
@@ -842,7 +877,7 @@ - (IBAction)play:(id)sender {
 				synth->setEndHandler(ROOT::evt::EventHandlerPtr(new ROOT::EndHandler(self)), 0);
 				synth->setBeatHandler(ROOT::evt::EventHandlerPtr(new ROOT::BeatHandler(self)), 0);
 				synth->setup(playData);
-				double delayInSeconds = 2.0;
+				double delayInSeconds = 0.0;
```
Comment (Redhairing) [inline Ivory/SSSampleViewController.mm]: We should probably document this delay variable more clearly

File: Ivory/IvoryKit/IvoryKit/AudioController.swift [hunk 6]
```diff
@@ -33,8 +33,10 @@ class AudioController:NSObject {
     
     var inputSource:InputSourceType
     var noteIntervalMap:NoteIntervalMapProtocol?
+    var ssVC:SSSampleViewController?
 
     var playerNode:AVAudioPlayerNode?
+    var acousticNode:AVAudioInputNode?
```
Comment (cnnrdltn) [inline Ivory/IvoryKit/IvoryKit/AudioController.swift]: changes in this file are for adding microphone input.  we have a "state machine" for different inputs and transitions between them setup or teardown the AVAudioEngine graph.  other than that much of the tap and processing is refactored to allow code reuse

File: Ivory/SSSampleViewController.mm [hunk 7]
```diff
@@ -99,6 +99,35 @@ @interface SSSampleViewController ()
...
+    [colour getRed:&red green:&green blue:&blue alpha:&alpha];
+    color.r = red;
+    color.g = green;
+    color.b = blue;
+    color.a = alpha;
+
+    [self.sysScrollView colourNotes:noteSet colour:color];
+}
+
+
```
Comment (Redhairing) [inline Ivory/SSSampleViewController.mm]: This added section might also need some documentation

File: Ivory/IvoryKit/IvoryKit/AudioController.swift [hunk 8]
```diff
@@ -145,18 +166,76 @@ class AudioController:NSObject {
...
     }
+    
+    func handleTapData(audioProcessor:AudioProcessor, buf:AVAudioPCMBuffer,when:AVAudioTime) -> (){
+        if self.startTime == nil {
+            self.startTime = AVAudioTime.secondsForHostTime(when.hostTime)
+        }
+        // time may be offset based on current time.  only want the difference from the starting time for the song
+        var time = AVAudioTime.secondsForHostTime(when.hostTime)
+        time -= self.startTime!
+
```
Comment (cnnrdltn) [inline Ivory/IvoryKit/IvoryKit/AudioController.swift]: need more comments

File: Ivory/IvoryKit/IvoryKit/AudioController.swift [hunk 9]
```diff
@@ -145,18 +166,76 @@ class AudioController:NSObject {
...
+                case .Omitted: missedNotes.append(ssNote)
+                case .ExtraNote:
+                    assert(false,"not implemented")
+                    break
+            }
+            }
+        }
+
+        if let ssvc = self.ssVC {
+            println(missedNotes)
```
Comment (cnnrdltn) [inline Ivory/IvoryKit/IvoryKit/AudioController.swift]: kill prints

File: Ivory/IvoryKit/IvoryKit/AudioController.swift [hunk 10]
```diff
@@ -145,18 +166,76 @@ class AudioController:NSObject {
...
+            println(correctNotes)
+            println(lateNotes)
+    
+            ssvc.colourNotes(missedNotes, colour: missedColor)
+            ssvc.colourNotes(correctNotes, colour: correctColor)
+            ssvc.colourNotes(lateNotes, colour: lateColor)
+        }
+
+    }
+
```
Comment (cnnrdltn) [inline Ivory/IvoryKit/IvoryKit/AudioController.swift]: good, refactor looks nice

File: Ivory/AudioProcessor.swift [hunk 11]
```diff
@@ -61,44 +61,36 @@ class AudioProcessor: NSObject {
...
+                    if find(computedValues, value.pitch) != nil {
+                        appearanceCount = appearanceCount + 1
+                        out[value] = ETNoteEvaluationResultType.OnTime
+                    } else {
+                        out[value] = ETNoteEvaluationResultType.Omitted
+                    }
+                }
+                println("\(appearanceCount) of \(expectedNotesLength) notes found")
+            }
+            return out
```
Comment (Redhairing) [inline Ivory/AudioProcessor.swift]: Return out? This should be named better - also the commented out return value below?

File: Ivory/IvoryKit/IvoryKit/AudioController.swift [hunk 12]
```diff
@@ -177,11 +256,27 @@ class AudioController:NSObject {
...
+            return inputNode
+        }
+    }
+    
     // MARK: Custom Init
     
     override init() {
         //TODO: Make sure setup GRAPH FOR THIS ON IT
-        // #warning WARNING: that could be an issue
-        self.inputSource = .AudioFile
```
Comment (cnnrdltn) [inline Ivory/IvoryKit/IvoryKit/AudioController.swift]: what do these warnings mean?

Comment (cnnrdltn) [inline Ivory/IvoryKit/IvoryKit/AudioController.swift]: @Redhairing @inkybink  - note for now this is where we decide if we want microphone or audio file for input.

File: Ivory/IvoryKit/PWNoteIntervalMap.swift [hunk 13]
```diff
@@ -54,13 +59,21 @@ class PWNoteIntervalMap: NSObject, NoteIntervalMapProtocol {
         return PWNoteInterval(start:Double(startTime), end:Double(endTime), content:pitch)
     }
 
-    func expectedNotesAt(startTime:Double, duration:Double) -> [Note] {
+    class func intervalFromSeeScoreNote(note:SeeScoreNote) -> PWNoteInterval {
```
Comment (cnnrdltn) [inline Ivory/IvoryKit/PWNoteIntervalMap.swift]: rename note as ssNote

File: Ivory/Note.swift [hunk 14]
```diff
@@ -272,5 +273,21 @@ class Note: NSObject {
 		self.pitch = pitch
 		super.init()
 	}
+}
+
+class SeeScoreNote: Note {
```
Comment (cnnrdltn) [inline Ivory/Note.swift]: container for swift wrapper (SSPlayDataWrapper).  new subclass of note needed for whenever would need to use SeeScore Play Data in Swift (maybe this should just replace Note and the notePlaydata member is an optional?)

add comments

File: Ivory/AudioProcessor.swift (hunk 1 above)
Comment (cnnrdltn) [inline Ivory/AudioProcessor.swift]: we're leaving sample size for now.  need duration as variable above.

File: Ivory/IvoryKit/IvoryKit/AudioController.swift (hunk 4 above)
Comment (cnnrdltn) [inline Ivory/IvoryKit/IvoryKit/AudioController.swift]: swift thing, needed for wrapping a c style array and exposing it to swift.  basically raw memory access.  have to use for realtime audio using avaudioengine

File: Ivory/SSSampleViewController.mm [hunk 15]
```diff
@@ -99,6 +99,35 @@ @interface SSSampleViewController ()
...
+//    [PartNote alloc] ini
+    
+    ssc::evt::NoteSet_t noteSet;
+    // assume only 1 part
+    int numparts = 1;
+    for (int pindex = 0; pindex < /*m_pd->*/numparts; ++pindex)
+    {
+        for (NSObject *note in notes) {
+            SSSampleAppDelegate *appDel = [[UIApplication sharedApplication] delegate];
+            sscore_pd_note playdata = [appDel getSeeScoreNote:note].sscore_pd_note;
```
Comment (cnnrdltn) [inline Ivory/SSSampleViewController.mm]: name is wrong "getSeeScoreNote"

**Pull request merged**